File: gfg/POTD/2025-02-06/solution.py

'''
# Node class
class Node:
    def __init__(self,val):
        self.data = val
        self.right = None
        self.left = None

'''
# Note: Build tree and return root node
class Solution:
    def buildTreeHelper(self, inorder, preorder, preIndex, s, e):
        if s > e:
            return None

        root = Node(preorder[preIndex[0]])
        preIndex[0] += 1

        index = inorder.index(root.data, s, e + 1)

        root.left = self.buildTreeHelper(inorder, preorder, preIndex, s, index - 1)
        root.right = self.buildTreeHelper(inorder, preorder, preIndex, index + 1, e)

        return root

    def buildTree(self, inorder, preorder):
        preIndex = [0]
        return self.buildTreeHelper(inorder, preorder, preIndex, 0, len(inorder) - 1)

File: gfg/POTD/2025-02-06/solution.java

/*
class Node {
    int data;
    Node left, right;

    Node(int key) {
        data = key;
        left = right = null;
    }
}
*/


class Solution {
    private static int searchValue(int[] inorder, int value, int s, int e) {
        for (int i = s; i <= e; i++) {
            if (inorder[i] == value) {
                return i;
            }
        }
        return -1;
    }

    private static Node buildTreeHelper(int[] inorder, int[] preorder, int[] preIndex, int s, int e) {
        if (s > e) {
            return null;
        }

        Node root = new Node(preorder[preIndex[0]]);
        preIndex[0]++;

        int index = searchValue(inorder, root.data, s, e);

        root.left = buildTreeHelper(inorder, preorder, preIndex, s, index - 1);
        root.right = buildTreeHelper(inorder, preorder, preIndex, index + 1, e);

        return root;
    }

    public static Node buildTree(int[] inorder, int[] preorder) {
        int[] preIndex = {0};
        return buildTreeHelper(inorder, preorder, preIndex, 0, inorder.length - 1);
    }
}
